### ISSUE TITLE
grpc module is not registering protobuf nested messages

### ISSUE THREAD

**Issue by thiagodpf** [grafana/k6#3160]
Hey folks!

I noticed that modules `k6/net/grpc` and `k6/experimental/grpc` are not registering the [nested types](https://protobuf.dev/programming-guides/proto3/#nested) defined in my `.proto` file. This means that when receiving the response (via stream) from the server, it is not possible to generate a javascript object via marshaling (and thus I am not able to proceed with the load test).

I've been debugging the code and saw that the fix could be made at this point (based on the module `k6/experimental/grpc`):
https://github.com/grafana/k6/blob/6137c5e2503389f8d205ea51ce269b71258cc842/vendor/github.com/grafana/xk6-grpc/grpc/client.go#L285-L295

Note that the above algorithm only goes through the _first level_ `messages` of the `.proto` file, but it is possible that a message has a list of _nested_ messages and so on recursively. It is possible to adjust the algorithm to traverse the data structure (tree) without the need to perform a recursion, although a recursion would be much simpler, since nobody defines so many nested levels.

Do you already have this fix mapped out to be resolved or can I propose a Pull Request?

**Comment by mstoykov** [comment]
Hi @thiagodpf, thanks for reporting this :bow: 

This seems like a big miss on our side, at least at first glance.

Are nested types used often? Are they a new feature, by any chance? This code is also a little [over a year old](https://github.com/grafana/k6/pull/2523) which seems like this is just not used all that much :thinking: ?

I am mostly asking to know just how likely is we have broken something instead of it just not working ever and nobody has been using it. I expect that later is the case, so I will mark this is an enhancement.


>  Do you already have this fix mapped out to be resolved or can I propose a Pull Request?

I don't think someone has reported this before or that we have started a fix. A PR is welcome here, if you want to do it.

I would probably propose doing it against the [xk6-grpc repo](https://github.com/grafana/xk6-grpc/), and we can transfer it in standard after that (potentially even in the same release).
Also, please add tests for this behaviour. 

Thanks again for the reporting this :bow:

**Comment by mstoykov** [comment]
I am not certain how you got the logged line you got.

Looking at the code 
https://github.com/grafana/xk6-grpc/blob/6095be08a1f0c2ef484a39b94a14c279bde1c393/grpc/stream.go#L176-L198

It seems to me you eitehr should have `error while reading from the stream` or `stream is cancelled/finished'. 

cc @olegbespalov 

>  Do you have a preference for the type of algorithm for traversing nested types? 

I don't think it matters so much. I do feel like having it be none recursive is actually going to be just as much code, but I propose you just do it, and we can figure out if it makes sense. 

> also because so many levels of nested types in .proto files are not common...

I have no idea - if we expect no more then 2-3 levels and at 10 we have practically no examples :+1: .

I am not certain at what level this will be a  problem either way, as I expect reading and parsing such a deeply neseted proto buf file is likely going to be more problematic then our handling of it :shrug:

**Comment by thiagodpf** [comment]
:small_blue_diamond: About this:

> _I am not certain how you got the logged line you got._

I only got that log because I was debugging the code to understand what was going on. Actually I forced that log in the terminal  debugging through VSCode: `call errConv.Error()`. :sweat_smile:


If I run k6 in `--verbose` mode, the log I would get would look like this:
```
DEBU[0005] error while reading from the stream           error="failed to marshal the message: proto:\u00a0google.protobuf.Any: unable to resolve \"type.googleapis.com/google.rpc.BadRequest.FieldViolation\": not found"
```


:small_blue_diamond: About algorithm to traverse nested message structure, I will propose something tonight :v::grin:

**Comment by thiagodpf** [comment]
@mstoykov I believe that I will need your guidance to proceed with the tests, in order to be able to reproduce (in the [xk6-grpc](https://github.com/grafana/xk6-grpc/tree/main) repo) the problem I reported.

**TL;DR** Could you tell me if there are test cases in which the process that runs the gRPC server is different from the process that runs the client? It seems to me that this code [here](https://github.com/grafana/xk6-grpc/tree/main/examples/grpc_server) (which represents the gRPC server) should run separately from the client, right? Could I run a test case that makes use of this server?

---

_Long story, go get a coffee and clear your mind to try to understand the problem..._

So that I can reproduce the error I'm having, I need to find a test scenario that takes me closer to that of real k6, where I only need to inform the `.proto` files and k6 loads them, without the code generated from the `.proto`. 
In practice, k6 does not have access to the code generated from the proto file, we just inform (in the test scripts) where the `.proto` file is located and k6 knows how to load them (and register de messages via `protoregistry.GlobalTypes` [here](https://github.com/grafana/xk6-grpc/blob/main/grpc/client.go#L290)).

I think it will be difficult for me to explain this (and I already see myself repeating the same thing), but... 😅
for the problem to be reproducible in a test environment, the code generated from the `.proto` cannot be part of the "k6" binary, that is, I need a test scenario in which the code generated from .proto is only on the server side, and k6 only has access to the .proto file in fact. In short, I need a way to separate the server side from the client side into different binaries.
this because there are `init()` functions in the "proto-generated" code that are executed as soon as the binary is loaded (this behavior is part of the golang package definition).

Here's an `init()` [example](https://github.com/grafana/xk6-grpc/blob/main/grpc/testutils/grpcservice/route_guide.pb.go#L444) of what I'm referring to!

If this `init()` above is executed on the client side (i.e. on K6) it does not represent the actual use of the software.
This means that this [line of code](https://github.com/grafana/xk6-grpc/blob/main/grpc/client.go#L290) (client side) is never executed in test environment (but in production it does!!!), as the `if` condition (`protoregistry.NotFound`) will never be `true` (because the damn `init()` already registered the message types in the `protoregistry.GlobalTypes`).

For it to be `true` (that is: the message types have _not yet_ been registered in the `protoregistry.GlobalTypes`), it is necessary that the binaries of the "proto-generated" code from the `.proto` file not be part of the client binary (i.e. k6), they should only stay on the server. 

I'm sorry for being so long-winded, but do you get the idea? 
Are there any test cases that take me the closest to what I described above?

**Comment by thiagodpf** [comment]
just to put you in context, I tried running [this test file](https://github.com/grafana/xk6-grpc/blob/main/grpc/tests/cmd_run_test.go) and it didn't work because it runs the client and server on the same binary.

**Comment by mstoykov** [comment]
@thiagodpf good catch :bow: - yeah, turns out that the tests for this functionality don't really do anything at this point.

After a quick internal discussion we do think that:
1. we should fix it - possibly with having a separate CI action to do this. Which has the problems that `go test ./...` will not catch it :(
2. we likely won't prioritize it high enough right away
3. we likely will first do it in `xk6-grpc` repo
4. we don't want this to block you

So our proposal is that:
1. you go ahead and do the code.
2. preferably still add unit tests that we will later fix (you can probably skip that for the first iteration - we might never user it and just use the next point)
3. make a test server and client scripts as the ones in the `examples` folder that only work with the fix.

This way we can get the actual fix and later figure out the exact way to test it going forward. 

You are also welcome to actually contribute a way to run the tests in separate processes.

**Comment by mstoykov** [comment]
@thiagodpf I should've probably written you about this, but the `xk6-grpc` depends on you having vendored the dependencies. I did open an issue before responding even https://github.com/grafana/xk6-grpc/issues/21.

So you can just run `go mod vendor` locally - we do this in the CI already.

Sorry for the inconvenience.

**Comment by thiagodpf** [comment]
Very good tip! :tada:
I was able to better understand the unit tests after running the command. It helped a lot!

I proposed a [PR](https://github.com/grafana/xk6-grpc/pull/23/files) on that repository, check it out when you can. If you want to debug to understand the algorithm, it's very easy, basically it's a traversal of a tree structure.

As for contributing ways to run the tests in separate processes, I'm afraid I'm not that familiar with github's CI :sweat_smile:. But thank you for opening a point of discussion on the subject. :raised_hands:
